### DISCUSSION TITLE
CDC-NCM device class unable to handle NTB size > 512B && Unable to use NCM with Linux HOST

### DISCUSSION THREAD

**Discussion by joshikeval3131** [hathach/tinyusb#2009] [Q&A]
Hi all, 

I am trying to use CDC-NCM dev class with Linux Host. Following problems were found during the integration. 

1. NCM dev class is unable to handle NTB of size > 512B i.e any NTB transfer which requires more than 1 OUT transfers. 

```
static void handle_incoming_datagram(uint32_t len)
{
  uint32_t size = len;

  if (len == 0) {
    return;
  }

  TU_ASSERT(size >= sizeof(nth16_t), );

  const nth16_t *hdr = (const nth16_t *)receive_ntb;
  TU_ASSERT(hdr->dwSignature == NTH16_SIGNATURE, );
  TU_ASSERT(hdr->wNdpIndex >= sizeof(nth16_t) && (hdr->wNdpIndex + sizeof(ndp16_t)) <= len, );

  const ndp16_t *ndp = (const ndp16_t *)(receive_ntb + hdr->wNdpIndex);
  TU_ASSERT(ndp->dwSignature == NDP16_SIGNATURE_NCM0 || ndp->dwSignature == NDP16_SIGNATURE_NCM1, );
  TU_ASSERT(hdr->wNdpIndex + ndp->wLength <= len, );

  int num_datagrams = (ndp->wLength - 12) / 4;
  ncm_interface.current_datagram_index = 0;
  ncm_interface.num_datagrams = 0;
  ncm_interface.ndp = ndp;
  for (int i = 0; i < num_datagrams && ndp->datagram[i].wDatagramIndex && ndp->datagram[i].wDatagramLength; i++)
  {
    ncm_interface.num_datagrams++;
  }
  tud_network_recv_renew();
}
```
Consider **handle_incoming_datagram** ApI, once it receives 1st OUT transaction it assumes that the datagram transfer is completed and call **tud_network_recv_renew()** function which will trigger another OUT transaction. 
 
![tusb_bug](https://user-images.githubusercontent.com/60086369/230072675-ec10adfd-91cc-4b6f-9f00-14257b0b39f1.png)

The next 23 bytes OUT transaction will reach to **handle_incoming_datagram** API and TU_ASSERT will fail because NTH signature won't match. 

**Accumulating the data of size  wBlockLength and then transferring it Lwip solves the issue.** 

2. There is a bug that has been identified in the Linux Host NCM implementation [Bug Link ](https://www.spinics.net/lists/linux-usb/msg239819.html) . Linux Host sends **SET_ETHERNET_PACKET_FILTER** request irrespective of disabling this capability in descriptor. 

TinyUSB NCM class driver as of now doesn't have **SET_ETHERNET_PACKET_FILTER** handling implemented. Thus it fails with Linux HOST PC. 

**Does it make sense to add this capability in dev class driver and give a callback which will set the filter on the networking Port. ? 
If yes , then I can raise a PR for the same.**

Thanks & Regards, 
K.Joshi

**Comment by kaushalmp** [comment]
reference to kernel Bug : [Bug 217290](https://bugzilla.kernel.org/show_bug.cgi?id=217290)

**Comment by joshikeval3131** [comment]
@majbthrd , 

Can you please comment on this ? 
As I can see that you were the reviewer of the NCM origin PR. 

Thanks

**Comment by joshikeval3131** [comment]
This issue has been resolved
